persist/sqlite/v2transactions.go

```diff
@@ -371,6 +352,75 @@ ORDER BY ts.transaction_order ASC`)
 	return nil
 }
 
+// fillV2TransactionFileContractRevisions fills in the file contracts for each
+// transaction.
+func fillV2TransactionFileContractRevisions(tx *txn, dbIDs []int64, txns []explorer.V2Transaction) error {
+	parentStmt, err := tx.Prepare(`SELECT fc.transaction_id, rev.confirmation_index, rev.confirmation_transaction_id, rev.resolution, rev.resolution_index, rev.resolution_transaction_id, fc.contract_id, fc.leaf_index, fc.capacity, fc.filesize, fc.file_merkle_root, fc.proof_height, fc.expiration_height, fc.renter_output_address, fc.renter_output_value, fc.host_output_address, fc.host_output_value, fc.missed_host_value, fc.total_collateral, fc.renter_public_key, fc.host_public_key, fc.revision_number, fc.renter_signature, fc.host_signature
+FROM v2_file_contract_elements fc
+INNER JOIN v2_transaction_file_contract_revisions ts ON (ts.parent_contract_id = fc.id)
+INNER JOIN v2_last_contract_revision rev ON (rev.contract_id = fc.contract_id)
+WHERE ts.transaction_id = ?
+ORDER BY ts.transaction_order ASC`)
+	if err != nil {
+		return fmt.Errorf("failed to prepare file contracts parent statement: %w", err)
+	}
+	defer parentStmt.Close()
+
+	revisionStmt, err := tx.Prepare(`SELECT fc.transaction_id, rev.confirmation_index, rev.confirmation_transaction_id, rev.resolution, rev.resolution_index, rev.resolution_transaction_id, fc.contract_id, fc.leaf_index, fc.capacity, fc.filesize, fc.file_merkle_root, fc.proof_height, fc.expiration_height, fc.renter_output_address, fc.renter_output_value, fc.host_output_address, fc.host_output_value, fc.missed_host_value, fc.total_collateral, fc.renter_public_key, fc.host_public_key, fc.revision_number, fc.renter_signature, fc.host_signature
+FROM v2_file_contract_elements fc
+INNER JOIN v2_transaction_file_contract_revisions ts ON (ts.revision_contract_id = fc.id)
+INNER JOIN v2_last_contract_revision rev ON (rev.contract_id = fc.contract_id)
+WHERE ts.transaction_id = ?
+ORDER BY ts.transaction_order ASC`)
+	if err != nil {
+		return fmt.Errorf("failed to prepare file contracts revision statement: %w", err)
+	}
+	defer revisionStmt.Close()
+
+	collectFileContracts := func(stmt *stmt, dbID int64) ([]explorer.V2FileContract, error) {
+		rows, err := stmt.Query(dbID)
+		if err != nil {
+			return nil, fmt.Errorf("failed to query file contracts: %w", err)
+		}
+		defer rows.Close()
+
+		var contracts []explorer.V2FileContract
+		for rows.Next() {
+			fce, err := scanV2FileContract(rows)
+			if err != nil {
+				return nil, fmt.Errorf("failed to scan file contract: %w", err)
+			}
+			contracts = append(contracts, fce)
+		}
+		return contracts, nil
+	}
+
+	for i, dbID := range dbIDs {
+		parents, err := collectFileContracts(parentStmt, dbID)
+		if err != nil {
+			return err
+		}
+
+		revisions, err := collectFileContracts(revisionStmt, dbID)
+		if err != nil {
+			return err
+		}
+
+		if len(parents) != len(revisions) {
+			panic("should have a parent for every v2 revision contract")
```

Comment:
Is this `panic` absolutely necessary? I can see how we want to, in some cases, panic when we have db corruption but seems quite harsh.

Reply:
I think the statements can be combined using `UNION ALL`, you add something like `"SELECT "parent", ... UNION ALL SELECT "revision"` to differentiate the two and then `collectFileContracts` returns two slices. 

```go
var kind string
if err := rows.Scan(&kind); err != nil {
    return nil, fmt.Errorf("failed to scan contract type: %w", err)
}

fce, err := scanV2FileContract(rows)
if err != nil {
    return nil, fmt.Errorf("failed to scan file contract: %w", err)
}

if kind == "parent" {
    parents = append(contracts, fce)
} else {
    revisions = append(contracts, fce)
}
```

Reply:
Not sure how well this would perform but something this would get rid of the duplicate select expression and clearly show what's happening.

```go
WITH 
    base_data AS (SELECT field1, field2, field3 FROM t1 INNER JOIN t2),
    revision_data AS (SELECT *, "revision" as kind FROM base_data),
    parent_data AS (SELECT *, "parent" as kind FROM base_data)
SELECT * FROM revision_data ORDER BY x
UNION ALL
SELECT * FROM parent_data ORDER BY x
```

Reply:
Tried doing something along these lines but I don't think this is worth it unless we find we're having performance issues with the current queries.  It makes the query more complicated and it seems that calling Scan twice on the same Rows doesn't work so we'd need to change the FCE scan helper too because we can't get the `kind` separately from the rest of the data.